// @enitoni/gears/build/bot/classes/ClientAdapter.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
const classes_1 = require("../../core/classes");
/**
 * Adapts a messaging interface so a [[Bot]] can understand it
 * @example
 * class Adapter extends ClientAdapter {
 *   protected register(options, hooks) {
 *     // Client refers to anything that exposes a messaging interface
 *     const client = new Client(options)
 *
 *     client.on("message", hooks.message)
 *     client.on("ready", hooks.ready)
 *     client.on("resume", hooks.ready)
 *     client.on("disconnect", hooks.unready)
 *     client.on("error", hooks.error)
 *
 *     return {
 *       client,
 *       methods: {
 *         start: async () => client.start(),
 *         getMessageContent: (message) => message
 *       }
 *     }
 *   }
 * }
 *
 * const adapter = new Adapter(options)
 * const bot = new Bot({ adapter, ... })
 * @template CO Client options
 * @template C Client
 * @template M Message
 * @category Bot
 */
class ClientAdapter extends classes_1.Emitter {
    constructor(options) {
        super();
        const { client, methods } = this.register(options, {
            message: message => this.emit("message", message),
            ready: () => this.emit("ready", undefined),
            unready: () => this.emit("unready", undefined),
            error: error => this.emit("error", error),
        });
        this.client = client;
        this.methods = methods;
    }
}
exports.ClientAdapter = ClientAdapter;
