// @enitoni/gears-discordjs/build/index.js
"use strict";
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : new P(function (resolve) { resolve(result.value); }).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
var __rest = (this && this.__rest) || function (s, e) {
    var t = {};
    for (var p in s) if (Object.prototype.hasOwnProperty.call(s, p) && e.indexOf(p) < 0)
        t[p] = s[p];
    if (s != null && typeof Object.getOwnPropertySymbols === "function")
        for (var i = 0, p = Object.getOwnPropertySymbols(s); i < p.length; i++) if (e.indexOf(p[i]) < 0)
            t[p[i]] = s[p[i]];
    return t;
};
function __export(m) {
    for (var p in m) if (!exports.hasOwnProperty(p)) exports[p] = m[p];
}
var __importStar = (this && this.__importStar) || function (mod) {
    if (mod && mod.__esModule) return mod;
    var result = {};
    if (mod != null) for (var k in mod) if (Object.hasOwnProperty.call(mod, k)) result[k] = mod[k];
    result["default"] = mod;
    return result;
};
Object.defineProperty(exports, "__esModule", { value: true });
const Gears = __importStar(require("@enitoni/gears"));
const discord_js_1 = require("discord.js");
class Adapter extends Gears.ClientAdapter {
    register(options, hooks) {
        const { token, listenToSelf } = options, clientOptions = __rest(options, ["token", "listenToSelf"]);
        const client = new discord_js_1.Client(clientOptions);
        client.on("message", message => {
            if (client.user.id !== message.author.id || listenToSelf) {
                hooks.message(message);
            }
        });
        client.on("ready", hooks.ready);
        client.on("resume", hooks.ready);
        client.on("disconnect", hooks.unready);
        client.on("error", hooks.error);
        return {
            client,
            methods: {
                start: () => __awaiter(this, void 0, void 0, function* () {
                    yield client.login(token);
                }),
                getMessageContent: (message) => message.content
            }
        };
    }
}
exports.Adapter = Adapter;
exports.Command = Gears.Command;
exports.CommandGroup = Gears.CommandGroup;
exports.CommandBuilder = Gears.CommandBuilder;
exports.CommandGroupBuilder = Gears.CommandGroupBuilder;
exports.Service = Gears.Service;
__export(require("./matchers"));
__export(require("./services"));
